src/Navigation/OnActionFactory.ts

```diff
@@ -13,30 +14,112 @@ import type {
   OnActionFactoryAttributes,
   BackNavigationAction,
   CancelFlowNavigationAction,
+  CloseNavigationAction,
   FinishFlowAndContinueNavigationAction,
   RequireConditionsNavigationAction,
   ValidateConditionsNavigationAction,
   NavigateNavigationAction,
+  CustomActionHandler,
+  StaticTreeNavigator,
 } from '../Model/Types'
+import {
+  findStaticNavigatorByStateKey,
+  transformForNearestExistingNavigator,
+} from '../Api/NavigationUtils'
+import { configManager } from '../Config'
 
 import { onBackAction } from './Back'
 import {
   onCancelFlowAction,
   onFinishFlowAndContinueAction,
 } from './Flow'
+import { onCloseAction } from './Close'
 import { onRequireConditionsAction } from './RequireConditions'
 import { onValidateConditionsAction } from './ValidateConditions'
 import { onNavigateAction } from './Navigate'
 
 const log = new Log('txo.conditional-navigation-react.Navigation.onActionFactory')
 
+const getHandlerWithPathFromParents = (staticTreeNavigator: StaticTreeNavigator, actionType: string): CustomActionHandler | undefined => {
+  const handler = staticTreeNavigator.handlerMap?.[actionType]
+  if (handler != null) {
+    return handler
+  }
+  const parent = staticTreeNavigator.getParent()
+  if (parent == null) {
+    return undefined
+  }
+  return getHandlerWithPathFromParents(parent, actionType)
+}
+
+const onExplicitNavigatorAction = (onActionAttributes: OnActionAttributes<NavigationAction & { navigatorId?: string }>): boolean => {
+  const {
+    action,
+    getState,
+    getRootState,
+    nextOnAction,
+    parentNavigationHelpers,
+    restArgs,
+  } = onActionAttributes
+
+  const { staticScreenTree } = configManager.config
+  const rootState = getRootState()
+  const navigatorState = getState()
+  const isRootNavigator = navigatorState.key === rootState.key
+  const staticNavigator = findStaticNavigatorByStateKey(staticScreenTree, rootState, navigatorState.key)
+  const navigatorId = staticNavigator?.id
+  const shouldConsumeAction = isNotEmptyString(action.navigatorId)
+    ? navigatorId === action.navigatorId || isRootNavigator
+    : true
+
+  const handler: CustomActionHandler | undefined = staticNavigator != null
```

Comment:
what if handler returns undefined for result, shouldn't we continue with other handler in parents?

Reply:
`getHandlerWithPathFromParents` gets handler recursively from all parents